File: services/api/varuna_api/routers/weather.py
"""Live weather for a city, proxied from Open-Meteo (TECH_SPEC 3.1; task D-05).

Everything else VARUNA serves is a forecast it computed. This is the one number on the citizen
screen that comes from outside: what the sky is doing now, and how much rain the next four hours
are expected to bring. It exists because a reader who opens the dashboard on a dry afternoon needs
to see something true immediately, before any baked run means anything to them.

**Why a proxy and not a browser fetch.** Three reasons, in order: the licence requires attribution
and a server can carry it in the payload rather than trusting each client to print it; one process
cache keeps the whole demo inside the free tier's 10,000 calls a day however many people load the
page; and the browser cannot fall back to a copy on our disk when the venue's network is gone.

**What it will and will not claim.** Open-Meteo answers for the grid cell containing the point,
not the point, so the response carries both `point` and `grid_point` and the distance between
them. `fetched_at`, `age_s` and `stale` are on every response, including the fresh ones, so a
screen never has to guess whether it is showing live weather or the last copy from before the
network went. Nothing here is mixed into a VARUNA product: it is context beside the forecast, not
an input to it.

**Caching, three deep.** The in-process cache (15 minutes, modelled on `routers/verify.py`'s
sweep cache, ADR-0058) answers the second reader. The last good response is also written to
`data/cache/weather-<city>.json`, so a restart, an upstream outage or `VARUNA_OFFLINE=1` still
has something honest to serve - stamped with its real age and flagged `stale`. When there is no
copy at all, the refusal is the section 12 envelope naming the reason, never a fabricated sky.
"""

from __future__ import annotations

import json
import math
import os
import re
import threading
from datetime import datetime, timedelta, timezone
from pathlib import Path
from typing import Annotated, Any

import structlog
from fastapi import APIRouter, Query
from pydantic import BaseModel, Field
from varuna_schemas import net
from varuna_schemas.constants import IST
from varuna_schemas.models import CityConfig
from varuna_schemas.net import get_json
from varuna_schemas.paths import city_config_path, data_dir

from varuna_api.state import api_error

log = structlog.get_logger("varuna.api.weather")

router = APIRouter(prefix="/v1", tags=["weather"])

OPEN_METEO_URL = "https://api.open-meteo.com/v1/forecast"
SOURCE = "open-meteo"
SOURCE_URL = "https://open-meteo.com/"
LICENCE = "CC BY 4.0"
LICENCE_URL = "https://creativecommons.org/licenses/by/4.0/"
ATTRIBUTION = "Weather data by Open-Meteo.com (CC BY 4.0)"
"""The licence requires attribution wherever the data is shown. Every response carries it."""

TTL_S = 900.0
"""Fifteen minutes. Open-Meteo updates hourly and its `current` block every 15 minutes."""

HOURLY_STEPS = 4
REQUEST_HOURS = 6
"""Six asked for, four returned: the first steps can already be in the past (see `_hourly`)."""

TIMEOUT_S = 6.0
RETRIES = 1
"""One retry. A dashboard waiting on weather is worse than a dashboard saying it is stale."""

CURRENT_FIELDS = "temperature_2m,relative_humidity_2m,precipitation,weather_code,wind_speed_10m"
HOURLY_FIELDS = ("precipitation", "precipitation_probability")

_CITY_RE = re.compile(r"^[a-z][a-z0-9_-]*$")

# WMO 4677 present-weather codes as Open-Meteo groups them (https://open-meteo.com/en/docs).
# A label, not a judgement: the numbers are the source's and the words are the WMO table's.
WMO_LABELS: dict[int, str] = {
    0: "Clear sky",
    1: "Mainly clear",
    2: "Partly cloudy",
    3: "Overcast",
    45: "Fog",
    48: "Depositing rime fog",
    51: "Light drizzle",
    53: "Moderate drizzle",
    55: "Dense drizzle",
    56: "Light freezing drizzle",
    57: "Dense freezing drizzle",
    61: "Slight rain",
    63: "Moderate rain",
    65: "Heavy rain",
    66: "Light freezing rain",
    67: "Heavy freezing rain",
    71: "Slight snowfall",
    73: "Moderate snowfall",
    75: "Heavy snowfall",
    77: "Snow grains",
    80: "Slight rain showers",
    81: "Moderate rain showers",
    82: "Violent rain showers",
    85: "Slight snow showers",
    86: "Heavy snow showers",
    95: "Thunderstorm",
    96: "Thunderstorm with slight hail",
    99: "Thunderstorm with heavy hail",
}


class GeoPoint(BaseModel):
    """A WGS84 point."""

    lon: float
    lat: float


class WeatherNow(BaseModel):
    """Open-Meteo's `current` block, renamed to VARUNA's vocabulary and units."""

    ts: datetime = Field(description="Observation time, ISO 8601 with the city's offset.")
    temperature_c: float | None = None
    humidity_pct: float | None = None
    precipitation_mm: float | None = Field(
        default=None, description="Rain in the source's last interval (15 minutes), millimetres."
    )
    wind_kmh: float | None = None
    weather_code: int | None = Field(default=None, description="WMO 4677 present-weather code.")
    weather: str | None = Field(default=None, description="The WMO table's label for the code.")


class WeatherStep(BaseModel):
    """One hourly step of the short forecast."""

    ts: datetime
    precipitation_mm: float | None = None
    precipitation_probability_pct: float | None = None


class Weather(BaseModel):
    """Current conditions and the next four hours, with its provenance and its age."""

    city: str
    point: GeoPoint = Field(description="The city AOI centre this was asked for.")
    grid_point: GeoPoint = Field(description="The centre of the model cell that answered.")
    grid_offset_km: float = Field(description="Distance between the two, kilometres.")
    elevation_m: float | None = None
    current: WeatherNow
    hourly: list[WeatherStep]
    fetched_at: datetime = Field(description="When this copy was retrieved from Open-Meteo.")
    age_s: float = Field(description="Seconds since `fetched_at`, computed per request.")
    stale: bool = Field(description=f"True once `age_s` exceeds the {int(TTL_S)} s cache window.")
    ttl_s: float = TTL_S
    source: str = SOURCE
    source_url: str = SOURCE_URL
    licence: str = LICENCE
    licence_url: str = LICENCE_URL
    attribution: str = ATTRIBUTION
    notes: list[str] = Field(
        default_factory=list, description="Why this copy is what it is; printed by the UI."
    )


_cache: dict[str, dict[str, Any]] = {}
_lock = threading.Lock()


def clear_cache() -> None:
    """Forget every kept response (tests)."""
    with _lock:
        _cache.clear()


def cache_path(city: str) -> Path:
    """Last good response on disk: ``data/cache/weather-<city>.json``."""
    return Path(data_dir()) / "cache" / f"weather-{city}.json"


def _city(raw: str) -> str:
    city = raw.strip().lower()
    if not _CITY_RE.match(city):
        raise api_error(
            400,
            "bad_city",
            f"{raw!r} is not a city slug. Use a lower-case name such as mumbai or chennai.",
        )
    return city


def _config(city: str) -> CityConfig:
    """The city's build config; its AOI centre is the point we ask about (never a literal)."""
    path = city_config_path(city)
    if not path.is_file():
        raise api_error(
            404,
            "city_not_configured",
            f"No city config at services/city/configs/{city}.yaml, so there is no point to ask "
            "Open-Meteo about. Configured cities: " + ", ".join(_configured_cities()) + ".",
        )
    try:
        return CityConfig.from_yaml(path)
    except (OSError, ValueError) as error:
        raise api_error(
            500, "city_config_invalid", f"{path.name} could not be read: {error}"
        ) from error


def _configured_cities() -> list[str]:
    folder = city_config_path("mumbai").parent
    if not folder.is_dir():
        return []
    return sorted(p.stem for p in folder.glob("*.yaml"))


def _km(a: GeoPoint, b: GeoPoint) -> float:
    lon_km = 111.320 * math.cos(math.radians(a.lat))
    return math.hypot((b.lon - a.lon) * lon_km, (b.lat - a.lat) * 110.540)


def _at_offset(raw: str | None, offset_s: int) -> datetime | None:
    """Open-Meteo returns local wall-clock times plus the offset; CLAUDE.md 12 wants the offset on."""
    if not raw:
        return None
    try:
        parsed = datetime.fromisoformat(raw)
    except ValueError:
        return None
    if parsed.tzinfo is not None:
        return parsed
    return parsed.replace(tzinfo=_tz(offset_s))


def _tz(offset_s: int) -> timezone:
    """The source's own offset; IST when it is the expected +05:30, so the label reads "IST"."""
    return IST if offset_s == 19800 else timezone(timedelta(seconds=offset_s))


def _hourly(payload: dict[str, Any], offset_s: int, after: datetime | None) -> list[WeatherStep]:
    """The next :data:`HOURLY_STEPS` steps strictly after ``after``.

    Open-Meteo's `forecast_hours=n` counts from the top of the current hour, so the first step is
    usually already in the past. Six are requested and the ones that have happened are dropped,
    which is why "the next four hours" is the next four, not "this hour and three more".
    """
    block = payload.get("hourly") or {}
    times = block.get("time") or []
    rain = block.get("precipitation") or []
    prob = block.get("precipitation_probability") or []
    steps: list[WeatherStep] = []
    for index, raw in enumerate(times):
        ts = _at_offset(raw, offset_s)
        if ts is None or (after is not None and ts <= after):
            continue
        steps.append(
            WeatherStep(
                ts=ts,
                precipitation_mm=_number(rain, index),
                precipitation_probability_pct=_number(prob, index),
            )
        )
        if len(steps) == HOURLY_STEPS:
            break
    return steps


def _number(values: list[Any], index: int) -> float | None:
    if index >= len(values):
        return None
    value = values[index]
    return float(value) if isinstance(value, (int, float)) else None


def _fetch(city: str, config: CityConfig) -> dict[str, Any]:
    """One live call to Open-Meteo, rendered into :class:`Weather` as a JSON-ready dict."""
    lon, lat = config.bbox.center
    raw = get_json(
        OPEN_METEO_URL,
        {
            "latitude": round(lat, 4),
            "longitude": round(lon, 4),
            "current": CURRENT_FIELDS,
            "hourly": list(HOURLY_FIELDS),
            "forecast_hours": REQUEST_HOURS,
            "timezone": "Asia/Kolkata",
        },
        timeout=TIMEOUT_S,
        retries=RETRIES,
    )
    if not isinstance(raw, dict) or "current" not in raw:
        msg = "Open-Meteo answered without a `current` block"
        raise net.NetworkError(msg)

    offset_s = int(raw.get("utc_offset_seconds") or 19800)
    current = raw.get("current") or {}
    now_ts = _at_offset(current.get("time"), offset_s) or datetime.now(IST)
    code = current.get("weather_code")
    code_int = int(code) if isinstance(code, (int, float)) else None
    point = GeoPoint(lon=round(lon, 6), lat=round(lat, 6))
    grid = GeoPoint(
        lon=float(raw.get("longitude", lon)),
        lat=float(raw.get("latitude", lat)),
    )
    weather = Weather(
        city=city,
        point=point,
        grid_point=grid,
        grid_offset_km=round(_km(point, grid), 2),
        elevation_m=_as_float(raw.get("elevation")),
        current=WeatherNow(
            ts=now_ts,
            temperature_c=_as_float(current.get("temperature_2m")),
            humidity_pct=_as_float(current.get("relative_humidity_2m")),
            precipitation_mm=_as_float(current.get("precipitation")),
            wind_kmh=_as_float(current.get("wind_speed_10m")),
            weather_code=code_int,
            weather=WMO_LABELS.get(code_int) if code_int is not None else None,
        ),
        hourly=_hourly(raw, offset_s, now_ts),
        fetched_at=datetime.now(IST),
        age_s=0.0,
        stale=False,
        notes=[],
    )
    return weather.model_dump(mode="json")


def _as_float(value: Any) -> float | None:
    return float(value) if isinstance(value, (int, float)) else None


def _write_disk(city: str, payload: dict[str, Any]) -> None:
    """Last good copy, written atomically so a crash mid-write cannot leave half a file."""
    path = cache_path(city)
    try:
        path.parent.mkdir(parents=True, exist_ok=True)
        tmp = path.with_suffix(".json.tmp")
        tmp.write_text(json.dumps(payload, indent=1), encoding="utf-8")
        os.replace(tmp, path)
    except OSError as error:  # a read-only volume must not fail the request
        log.warning("weather.cache_write_failed", city=city, error=str(error))


def _read_disk(city: str) -> dict[str, Any] | None:
    path = cache_path(city)
    try:
        payload = json.loads(path.read_text(encoding="utf-8"))
    except (OSError, ValueError):
        return None
    if isinstance(payload, dict) and payload.get("fetched_at"):
        return payload
    return None


def _cached(city: str) -> dict[str, Any] | None:
    """The newest copy this process can reach: memory first, then the last good file."""
    with _lock:
        hit = _cache.get(city)
    if hit is not None:
        return hit
    disk = _read_disk(city)
    if disk is not None:
        with _lock:
            _cache.setdefault(city, disk)
    return disk


def _store(city: str, payload: dict[str, Any]) -> None:
    with _lock:
        _cache[city] = payload
    _write_disk(city, payload)


def _age_s(payload: dict[str, Any]) -> float:
    fetched = _at_offset(str(payload.get("fetched_at")), 19800)
    if fetched is None:
        return float("inf")
    return max(0.0, (datetime.now(IST) - fetched).total_seconds())


def _stamp(payload: dict[str, Any], notes: list[str]) -> Weather:
    """Re-age a kept copy at serve time; the stored `age_s` is never the one returned."""
    age = _age_s(payload)
    body = {**payload, "age_s": round(age, 1), "stale": age > TTL_S}
    body["notes"] = [*payload.get("notes", []), *notes]
    return Weather.model_validate(body)


@router.get(
    "/weather",
    response_model=Weather,
    summary="Current conditions and the next four hours from Open-Meteo",
)
def weather(
    city: Annotated[str, Query(description="City slug, e.g. mumbai or chennai.")] = "mumbai",
) -> Weather:
    """Live weather for the city's AOI centre, cached for fifteen minutes.

    Degrades rather than fails: an upstream error, a timeout or `VARUNA_OFFLINE=1` serves the last
    good copy with its true age and `stale: true`. Only when no copy has ever been fetched does
    this answer 503, with the reason named.
    """
    slug = _city(city)
    config = _config(slug)
    kept = _cached(slug)

    if kept is not None and _age_s(kept) < TTL_S:
        return _stamp(kept, [])

    if net.offline():
        if kept is not None:
            return _stamp(
                kept,
                [
                    "VARUNA_OFFLINE=1, so Open-Meteo was not contacted. This is the last copy "
                    "fetched before the network was switched off."
                ],
            )
        raise api_error(
            503,
            "offline",
            "VARUNA_OFFLINE=1 blocks outbound requests and no weather has been cached for "
            f"{slug} yet. Unset VARUNA_OFFLINE, or run once with the network to fill "
            f"data/cache/weather-{slug}.json.",
        )

    try:
        fresh = _fetch(slug, config)
    except net.NetworkError as error:
        log.warning("weather.upstream_failed", city=slug, error=str(error))
        if kept is not None:
            return _stamp(
                kept,
                [f"Open-Meteo could not be reached ({error}); showing the last good copy."],
            )
        raise api_error(
            503,
            "weather_unavailable",
            f"Open-Meteo could not be reached ({error}) and nothing has been cached for {slug}. "
            "The flood forecast on this page does not depend on it.",
        ) from error

    _store(slug, fresh)
    log.info("weather.fetched", city=slug, steps=len(fresh.get("hourly", [])))
    return _stamp(fresh, [])


__all__ = [
    "ATTRIBUTION",
    "LICENCE",
    "OPEN_METEO_URL",
    "SOURCE",
    "TTL_S",
    "GeoPoint",
    "Weather",
    "WeatherNow",
    "WeatherStep",
    "cache_path",
    "clear_cache",
    "router",
]

File: apps/command/lib/api/weather.ts
/**
 * Live weather for a city (`GET /v1/weather`; TECH_SPEC 5, task D-13).
 *
 * **Why this client refuses to guess.** Every other number the dashboard prints comes out of a
 * baked run. This one comes from outside, and it is the only thing on the screen that is true
 * *now* - which makes it the one number a reader will trust without checking. So the loader has
 * exactly three outcomes and no fourth: a copy the API served (fresh or stale, with its real
 * age), or `unavailable` with the reason in words. A 404, a dead network and a refusing proxy all
 * land in `unavailable`; none of them produces a temperature. CLAUDE.md rule 6 says a number on
 * screen comes from an artifact or a labelled source, and "the last one I saw" is neither.
 *
 * **Staleness is the API's word, not ours.** `age_s` and `stale` are recomputed by the server on
 * every response against `fetched_at`, so this file never re-ages anything: it reports what it was
 * told. `ageLabel()` turns the seconds into the sentence UI_SPEC 5 asks for.
 *
 * **The grid caveat.** Open-Meteo answers for the model cell containing the point, not the point.
 * `grid_offset_km` is how far that cell's centre is from the city's AOI centre; past
 * {@link GRID_NOTICE_KM} the dialog says so rather than letting a reader think the thermometer is
 * in Dadar.
 */

import { apiFetch, errorMessage, isApiError } from "@/lib/api/client";

/** A WGS84 point, as the API serialises it. */
export interface WeatherPoint {
  lon: number;
  lat: number;
}

/** Current conditions. Every field is optional because the source may omit any of them. */
export interface WeatherNow {
  /** Observation time, ISO 8601 with the city's offset. */
  ts: string;
  temperatureC: number | null;
  humidityPct: number | null;
  /** Rain in the source's last interval, millimetres. */
  precipitationMm: number | null;
  windKmh: number | null;
  /** WMO 4677 present-weather code. */
  weatherCode: number | null;
  /** The WMO table's label for the code, e.g. "Moderate rain". */
  weather: string | null;
}

/** One hourly step of the short forecast. */
export interface WeatherStep {
  ts: string;
  precipitationMm: number | null;
  precipitationProbabilityPct: number | null;
}

/** A served response: the sky over the city, with its provenance and its age. */
export interface Weather {
  city: string;
  /** The city AOI centre the question was asked about. */
  point: WeatherPoint;
  /** The centre of the model cell that answered it. */
  gridPoint: WeatherPoint;
  /** Distance between the two, kilometres. */
  gridOffsetKm: number;
  elevationM: number | null;
  current: WeatherNow;
  hourly: WeatherStep[];
  /** When this copy was retrieved from the source. */
  fetchedAt: string;
  /** Seconds since `fetchedAt`, computed by the API for this request. */
  ageS: number;
  /** True once the copy is older than the API's cache window. */
  stale: boolean;
  ttlS: number;
  source: string;
  sourceUrl: string;
  licence: string;
  licenceUrl: string;
  /** The attribution the licence requires; printed verbatim wherever the data is shown. */
  attribution: string;
  /** Why this copy is what it is - offline, upstream down - in the API's own words. */
  notes: string[];
}

/** What the screen has: a served reading, or nothing and the reason. */
export type WeatherState =
  | { kind: "loading" }
  | { kind: "ready"; weather: Weather }
  | { kind: "unavailable"; reason: string };

/**
 * Past this far from the AOI centre the dialog says the answer is for a model grid cell.
 *
 * Two kilometres is roughly Hindmata to Dadar TT: near enough that "Mumbai" is honest, far enough
 * that a reader who knows the city would otherwise be entitled to read it as their street.
 */
export const GRID_NOTICE_KM = 2;

/** The dashboard's city until the switcher threads one through (task D-09). */
export const DEFAULT_WEATHER_CITY = "mumbai";

/**
 * How long a weather call is waited for.
 *
 * The API's own upstream timeout is 6 s with one retry, so a cold call that has to reach
 * Open-Meteo can legitimately take about 13 s before it either answers or falls back to its disk
 * copy. A shorter ceiling here would report "unavailable" over an API that was about to answer.
 */
export const WEATHER_TIMEOUT_MS = 20_000;

function num(value: unknown): number | null {
  return typeof value === "number" && Number.isFinite(value) ? value : null;
}

function point(raw: unknown): WeatherPoint {
  const record = (raw ?? {}) as Record<string, unknown>;
  return { lon: num(record.lon) ?? 0, lat: num(record.lat) ?? 0 };
}

function now(raw: unknown): WeatherNow {
  const record = (raw ?? {}) as Record<string, unknown>;
  return {
    ts: String(record.ts ?? ""),
    temperatureC: num(record.temperature_c),
    humidityPct: num(record.humidity_pct),
    precipitationMm: num(record.precipitation_mm),
    windKmh: num(record.wind_kmh),
    weatherCode: num(record.weather_code),
    weather: typeof record.weather === "string" ? record.weather : null,
  };
}

function step(raw: unknown): WeatherStep {
  const record = (raw ?? {}) as Record<string, unknown>;
  return {
    ts: String(record.ts ?? ""),
    precipitationMm: num(record.precipitation_mm),
    precipitationProbabilityPct: num(record.precipitation_probability_pct),
  };
}

/** The wire body as the screen's shape. Missing numbers stay null; they are never defaulted to 0. */
export function weatherFromBody(body: Record<string, unknown>): Weather {
  return {
    city: String(body.city ?? ""),
    point: point(body.point),
    gridPoint: point(body.grid_point),
    gridOffsetKm: num(body.grid_offset_km) ?? 0,
    elevationM: num(body.elevation_m),
    current: now(body.current),
    hourly: Array.isArray(body.hourly) ? body.hourly.map(step) : [],
    fetchedAt: String(body.fetched_at ?? ""),
    ageS: num(body.age_s) ?? 0,
    stale: body.stale === true,
    ttlS: num(body.ttl_s) ?? 900,
    source: String(body.source ?? ""),
    sourceUrl: String(body.source_url ?? ""),
    licence: String(body.licence ?? ""),
    licenceUrl: String(body.licence_url ?? ""),
    attribution: String(body.attribution ?? ""),
    notes: Array.isArray(body.notes) ? body.notes.map((n) => String(n)) : [],
  };
}

/**
 * Load the weather for `city`, or say why there is none.
 *
 * Never throws and never resolves to a partial reading: a failure of any kind - the API refusing,
 * the network gone, the body unreadable - becomes `unavailable` carrying the API's own sentence
 * where there is one, because CLAUDE.md 6.8 wants an error that names the fix rather than a shrug.
 */
export async function loadWeather(
  city = DEFAULT_WEATHER_CITY,
  options: { signal?: AbortSignal; timeoutMs?: number } = {},
): Promise<WeatherState> {
  try {
    const body = await apiFetch<Record<string, unknown>>("/v1/weather", {
      query: { city },
      timeoutMs: options.timeoutMs ?? WEATHER_TIMEOUT_MS,
      signal: options.signal,
    });
    return { kind: "ready", weather: weatherFromBody(body) };
  } catch (error) {
    if (isApiError(error) && error.code === "aborted") {
      return { kind: "unavailable", reason: "The weather request was cancelled." };
    }
    return {
      kind: "unavailable",
      reason: errorMessage(
        error,
        "The live weather source could not be reached. The flood forecast on this page does not depend on it.",
      ),
    };
  }
}

/**
 * "2 min ago", "just now", "41 min ago", "3 h ago" - the age UI_SPEC 5 prints beside the source.
 *
 * Rounded down, so a copy is never described as younger than it is.
 */
export function ageLabel(ageS: number): string {
  if (!Number.isFinite(ageS) || ageS < 0) return "age unknown";
  if (ageS < 60) return "just now";
  const minutes = Math.floor(ageS / 60);
  if (minutes < 60) return `${minutes} min ago`;
  const hours = Math.floor(minutes / 60);
  if (hours < 24) return `${hours} h ago`;
  const days = Math.floor(hours / 24);
  return `${days} d ago`;
}

/** "Open-Meteo · CC BY 4.0 · fetched 2 min ago" (UI_SPEC 5), built from what the API sent. */
export function sourceLine(weather: Weather): string {
  const name = weather.source === "open-meteo" ? "Open-Meteo" : weather.source;
  const parts = [name, weather.licence].filter((part) => part.length > 0);
  parts.push(`fetched ${ageLabel(weather.ageS)}`);
  return parts.join(" · ");
}

/** True when the answering model cell is far enough away that the dialog must say so. */
export function isGridDistant(weather: Weather): boolean {
  return Number.isFinite(weather.gridOffsetKm) && weather.gridOffsetKm > GRID_NOTICE_KM;
}

/** Temperature with its unit, or null when the source did not send one - never a zero. */
export function temperatureLabel(weather: Weather): string | null {
  const value = weather.current.temperatureC;
  return value === null ? null : `${Math.round(value)} °C`;
}
